components/script/dom/intersectionobserver.rs

```diff
@@ -271,6 +302,352 @@ impl IntersectionObserver {
             .borrow_mut()
             .retain(|element| &**element != target);
     }
+
+    /// <https://w3c.github.io/IntersectionObserver/#queue-an-intersectionobserverentry>
+    #[allow(clippy::too_many_arguments)]
+    fn queue_an_intersectionobserverentry(
+        &self,
+        document: &Document,
+        time: CrossProcessInstant,
+        root_bounds: Rect<Au>,
+        bounding_client_rect: Rect<Au>,
+        intersection_rect: Rect<Au>,
+        is_intersecting: bool,
+        is_visible: bool,
+        intersection_ratio: f64,
+        target: &Element,
+        can_gc: CanGc,
+    ) {
+        let rect_to_domrectreadonly = |rect: Rect<Au>| {
+            DOMRectReadOnly::new(
+                self.owner_doc.window().as_global_scope(),
+                None,
+                rect.origin.x.to_f64_px(),
+                rect.origin.y.to_f64_px(),
+                rect.size.width.to_f64_px(),
+                rect.size.height.to_f64_px(),
+                can_gc,
+            )
+        };
+
+        let root_bounds = rect_to_domrectreadonly(root_bounds);
+        let bounding_client_rect = rect_to_domrectreadonly(bounding_client_rect);
+        let intersection_rect = rect_to_domrectreadonly(intersection_rect);
+
+        // Step 1-2
+        // > 1. Construct an IntersectionObserverEntry, passing in time, rootBounds,
+        // >    boundingClientRect, intersectionRect, isIntersecting, and target.
+        // > 2. Append it to observer’s internal [[QueuedEntries]] slot.
+        self.queued_entries.borrow_mut().push(
+            IntersectionObserverEntry::new(
+                self.owner_doc.window(),
+                None,
+                document
+                    .owner_global()
+                    .performance()
+                    .to_dom_high_res_time_stamp(time),
+                Some(&root_bounds),
+                &bounding_client_rect,
+                &intersection_rect,
+                is_intersecting,
+                is_visible,
+                Finite::wrap(intersection_ratio),
+                target,
+                can_gc,
+            )
+            .as_traced(),
+        );
+        // > Step 3
+        // Queue an intersection observer task for document.
+        document.queue_an_intersection_observer_task();
+    }
+
+    /// Step 3.1-3.5 of <https://w3c.github.io/IntersectionObserver/#notify-intersection-observers-algo>
+    pub(crate) fn invoke_callback(&self) {
```

Comment:
We should rename this method, because it doesn't always invoke the callback.

Maybe `invoke_callback_if_necessary`?